[Hub Generated] Review request for Microsoft.Azure.Search.Service to add version 2017-11-11-preview

bitsforthought opened the pull request:
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/brjohnstmsft/azure-rest-api-specs/fix-search).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

bitsforthought committed:
[Azure Search] Make connectionString create-only to fix validation warning

This change is just to satisfy Swagger spec validation; It has no effect on
code generation (at least for C#).

bitsforthought committed:
[Azure Search] Fix various Swagger validation errors

This is our first major push towards making searchservice.json pass all
Swagger validation. The validator found several issues that this change
fixes:
  - "Extensible enums" were being modeled incorrectly. On the wire, they
    are just strings, but we were modeling them as objects with a "name"
    property. This reflected the generated code, but not the wire format.
    The solution is to model them as proper enums in the Swagger. This causes
    new validation errors, since these enums should be open-ended, but that's
    an issue with the validator not understanding "extensible enums"; it's
    not a problem with our spec per se.
  - Properties of Skill were marked as "required", when they were being
    omitted or set to null in examples. This is because these properties
    actually appear to be optional from the point of view of the REST API, so
    now we model them as such.
  - "documentdb" was deprecated as a DataSourceType, but not replaced with
    "cosmosdb" in the examples. This is now fixed.
  - The example for "Get Indexer Execution Status" included properties in the
    response that aren't modeled in the Swagger. These properties actually do
    appear on the wire, but they shouldn't and will be removed in a future API
    version. In the meantime, I have removed them from the example.

bitsforthought committed:
[Azure Search] Applying consistent JSON formatting to Swagger spec

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

brjohnstmsft commented:
@jianghaolu Contrary to the PR title, this is not a new API version. These changes are meant to address validation errors that we have never attempted to fix before. See individual commits for details. I have already requested exceptions for the remaining validation errors as I believe them to be limitations or bugs in the validator itself.

brjohnstmsft commented:
FYI @Yahnoosh 

Also FYI @arv100kri @mgottein since this contains Cognitive Search-related changes. See commits for details. A .NET SDK PR will follow once this is merged.

brjohnstmsft commented:
@jianghaolu Also, please ignore the auto-generated .NET SDK PR. Our SDK includes custom code and other scripts that run during code generation. The auto-generated PRs will probably never succeed for our .NET SDK. Ask @shahabhijeet if you need details.

arv100kri commented on specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/examples/SearchServiceCreateDataSource.json:
```diff
@@ -6,7 +6,7 @@
         "dataSource": {
             "name": "mydocdbdatasource",
             "description": "My Cosmos DB data source.",
-            "type": "documentdb",
```
Thanks!

arv100kri commented on specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/examples/SearchServiceCreateOrUpdateSkillset.json:
```diff
@@ -8,212 +8,272 @@
-				"name": "demoskillset",
-				"description": "Extract entities, detect language and extract key-phrases",
-				"skills": [{
 					"@odata.type": "#Microsoft.Skills.Text.NamedEntityRecognitionSkill",
```
Do you mind finding and replacing in the examples usage of "NamedEntityRecognitionSkill" to "EntityRecognitionSkill"... I must've forgot to do this as part of my swagger update where I added entity recognition skill

arv100kri replied:
The files that require this find and replace are
SearchServiceCreateOrUpdateSkillset.json
SearchServiceCreateSkillset.json
SearchserviceGetSkillset.json
SearchServiceListSkillsets.json

brjohnstmsft replied:
Sure

brjohnstmsft replied:
I attempted this, but it broke Swagger validation, so I rolled it back. It's up to you and your team to update this.

arv100kri commented:
Regarding this
>The example for "Get Indexer Execution Status" included properties in the
    response that aren't modeled in the Swagger. These properties actually do
    appear on the wire, but they shouldn't and will be removed in a future API
    version. In the meantime, I have removed them from the example.

@mgottein please comment on whether the comment is accurate or not

arv100kri commented on specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json:
```diff
@@ -4275,7 +4446,7 @@
         }
       },
       "required": [
-          "@odata.type", "description", "context", "inputs", "outputs"
```
So this is a bit weird... description and context are truly optional...

inputs and outputs are technically optional too, but if left as null for any skill (save the `WebApiSkill`) the API will throw a validation error... I am unsure how to model that in swagger...

brjohnstmsft replied:
Will fix; Thanks for confirming

arv100kri reviewed (approved):
Generally looks good, I had one request (NamedEntity -> Entity in examples) and the question I posed to @mgottein

brjohnstmsft commented:
> Regarding this
> 
> The example for "Get Indexer Execution Status" included properties in the
> response that aren't modeled in the Swagger. These properties actually do
> appear on the wire, but they shouldn't and will be removed in a future API
> version. In the meantime, I have removed them from the example.
> 
> @mgottein please comment on whether the comment is accurate or not

@arv100kri I am asserting that it is accurate. :) The Get Indexer Status API re-uses a model class from the Indexing API internally (IndexResult). This is an accident of implementation. That type includes "success" and "statusCode" flags which are meaningless for "ItemError" (errors never succeed, and status code is meaningless for at least some indexer error conditions). I assume that is why these properties are not modeled in the Swagger (it was probably Eugene who did this work, and I would have reviewed and approved it back then).

bitsforthought committed:
[Azure Search] Make inputs/outputs properties of Skill required

It turns out these properties are required by the API. Also replaced the
deprecated NER skill in examples.

bitsforthought committed:
[Azure Search] Standardize formatting of JSON examples

bitsforthought committed:
[Azure Search] Rolling back NamedEntityRecognition example update

bitsforthought committed:
[Azure Search] Make Swagger spec accurately reflect the semantics of Field

Before this change, the descriptions in the Swagger for the Field type did
not accurately describe the semantics of the Azure Search REST API,
particularly around what the default values of the searchable, filterable,
sortable, and facetable properties were. Instead, they reflected the defaults
of our custom implementation of Field in the .NET SDK.

This change corrects the descriptions to accurately reflect the REST API. It
also adds more detail in anticipation of generating REST API documentation
from the Swagger (which we don't currently do).

A few other changes to Field are related to code generation:
 - Removed the x-ms-external flag so that AutoRest can generate this type.
 - Removed x-nullable: false and x-ms-client-name since they again reflect
   only the custom .NET code and not the desired REST API semantics.
 - Added regex transform rules to readme.md so that we can keep some of our
   custom code for Field in .NET purely for backward compatibility purposes.
   Our intention going forward for other languages is to rely solely on code
   generation for the Field type.

brjohnstmsft commented:
@jianghaolu Can you review and merge this please?
